Examples/SubscriberExample/Sources/Utils/LocationAnimator/DefaultLocationAnimator.swift

```diff
@@ -0,0 +1,242 @@
+                    var step = step
+                    step.updateDuration(animationStepDuration)
+                    return step
+                }
+            
+                // Store last position from animation steps array
+                // In next iteration this position could be start position for the first step of the animation
+                self.previousFinalPosition = steps.last?.endPosition
+                
+                while !steps.isEmpty {
+                    self.animateStep(steps.removeFirst())
+                }
+            }
+        }.store(in: &subscriptions)
+    }
+   
+    func animateLocationUpdate(location: LocationUpdate, interval: Double) {
+        animationRequestSubject.send(AnimationRequest(locationUpdate: location, interval: interval))
+    }
+    
+    func trackablePosition(_ closure: @escaping (Position) -> Void) {
+        self.trackablePositionClosure = closure
+    }
+    
+    func fragmentaryPosition(_ closure: @escaping (Position) -> Void) {
+        self.fragmentaryPositionClosure = closure
+    }
+    
+    private func startAnimationLoop() {
+        let displayLink = CADisplayLink(target: self, selector: #selector(animationLoop))
+        displayLink.add(to: .current, forMode: .default)
+        
+        displayLinkStartTime = CFAbsoluteTimeGetCurrent()
+        self.displayLink = displayLink
+    }
+    
+    private func stopAnimationLoop() {
+        guard let displayLink = displayLink else {
+            return
+        }
+        
+        displayLink.remove(from: .current, forMode: .default)
+        displayLink.invalidate()
+    }
+    
+    private func createAnimationStepsFromRequest(_ request: AnimationRequest) -> [AnimationStep] {
+        let requestPositions = (request.locationUpdate.skippedLocations + [request.locationUpdate.location]).map { $0.toPosition() }
+        return requestPositions.enumerated().reduce([]) { [weak self] partialResult, value in
+            guard let self = self else {
+                return []
+            }
+            
+            let startPosition = value.offset == 0
+            ? self.getNewAnimationStartingPosition(locationUpdate: request.locationUpdate)
+            : requestPositions[value.offset - 1]
+            
+            return partialResult + [AnimationStep(startPosition: startPosition, endPosition: value.element, duration: self.unknownDuration)]
+        }
+    }
+    
+    private func getNewAnimationStartingPosition(locationUpdate: LocationUpdate) -> Position {
+        previousFinalPosition
+            ?? locationUpdate.skippedLocations.first?.toPosition()
+            ?? locationUpdate.location.toPosition()
+    }
+    
+    private func animateStep(_ step: AnimationStep?) {
+        guard let step = step, step.duration > .zero else {
+            return
+        }
+        
+        let displayLinkDuration = self.displayLink?.duration ?? self.defaultDisplayLinkDuration
```

Comment:
My knowledge here is a bit shaky, so please tell me if you think I might be wrong.

I believe that `CADisplayLink` does not necessarily fire at a fixed interval, but that, rather, it will only update each time the system renders a new frame. Which means, if the system is not able to run at its full frame rate for whatever reason, that your calculation of the number of frames that you need to render will be too high. And then, you'll proceed to animate all of these calculated frames anyway, but over a longer period of time than you'd intended to. This will introduce a lag between the location data being received and displayed on the map.

Would it perhaps be better to calculate the next frame position each time the `CADisplayLink` fires, using the `duration` to estimate the amount of time until the next time it fires?

Reply:
Sorry, I don't think you actually need to estimate anything - I think you just want to see how long has elapsed since the start of the animation.

Reply:
All I want is to calculate the number of keyframes (interpolated) between `startPosition` and `endPosition`. Since each keyframe should be running with `self.displayLink?.duration` interval, I guess I need still need this value.

Reply:
Ok, I understand what you mean. I guess you put a lot of effort into this review, and I appreciate it, but I won't change anything here. This implementation is just some idea of how to resolve the animation of the map annotation. It is one of many solutions, and I know it's not perfect, but I'm sure it's good enough as a solution "for free" for the community. 
Furthermore, it's based on the AAT Android Location Animator.

Reply:
> but I'm sure it's good enough as a solution "for free" for the community

I think that if we were only going to include this code in our example app, then I'd agree. But if it's going to go into the SDK too (which from #293 it seems it is) then personally I'd prefer to iron out these issues.

I'll follow your lead on this one though - would you like me to resolve?

Reply:
(I also don't have a sense of how much other work you've got on your plate and where this fits in priority-wise 🙂)

Reply:
Also, it's worth bearing in mind that on iOS devices with a variable refresh rate, [the refresh rate can change at any time](https://developer.apple.com/documentation/quartzcore/optimizing_promotion_refresh_rates_for_iphone_13_pro_and_ipad_pro):

> Custom animations in your app need to be able to adapt to changes in refresh rates. Display refresh rates can change for many reasons and your app must not presume any specific refresh rate, at any time. For example, the system disables faster refresh rates in low power mode or if a device gets hot. Also, while UIKit and Core Animation are managing various GUI elements, Core Animation might elect to vary the refresh rate to provide an enhanced user experience.

Reply:
check here f5be0b71084bc4bd90970bef1734713d0bdbb28a

Reply:
@lawrence-forooghian I appreciate for this very useful and productive conversation. I will create an issue to improve animation logic based on your comments and resolve this conversation in favour of moving this PR forward